File: app/src/main/java/com/electricdreams/shellshock/feature/items/ItemSelectionActivity.java
package com.electricdreams.shellshock.feature.items;

import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.electricdreams.shellshock.ModernPOSActivity;
import com.electricdreams.shellshock.R;
import com.electricdreams.shellshock.core.model.BasketItem;
import com.electricdreams.shellshock.core.model.Item;
import com.electricdreams.shellshock.core.util.BasketManager;
import com.electricdreams.shellshock.core.util.CurrencyManager;
import com.electricdreams.shellshock.core.util.ItemManager;
import com.electricdreams.shellshock.core.worker.BitcoinPriceWorker;

import java.io.File;
import java.util.Currency;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

public class ItemSelectionActivity extends AppCompatActivity {

    private ItemManager itemManager;
    private BasketManager basketManager;
    private BitcoinPriceWorker bitcoinPriceWorker;
    
    private RecyclerView recyclerView;
    private TextView emptyView;
    private TextView basketCountView;
    private TextView basketTotalView;
    private Button viewBasketButton;
    private Button checkoutButton;
    
    private ItemAdapter adapter;
    
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_item_selection);
        
        // Set up toolbar
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        
        // Initialize managers
        itemManager = ItemManager.getInstance(this);
        basketManager = BasketManager.getInstance();
        bitcoinPriceWorker = BitcoinPriceWorker.getInstance(this);
        
        // Initialize views
        recyclerView = findViewById(R.id.items_recycler_view);
        emptyView = findViewById(R.id.empty_view);
        basketCountView = findViewById(R.id.basket_count);
        basketTotalView = findViewById(R.id.basket_total);
        viewBasketButton = findViewById(R.id.view_basket_button);
        checkoutButton = findViewById(R.id.checkout_button);
        
        // Set up RecyclerView
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        adapter = new ItemAdapter(itemManager.getAllItems());
        recyclerView.setAdapter(adapter);
        
        // Set up empty view
        updateEmptyViewVisibility();
        
        // Set up buttons
        viewBasketButton.setOnClickListener(v -> {
            Intent intent = new Intent(ItemSelectionActivity.this, BasketActivity.class);
            startActivity(intent);
        });
        
        checkoutButton.setOnClickListener(v -> proceedToCheckout());
        
        // Update basket summary
        updateBasketSummary();
        
        // Make sure bitcoin price worker is running
        bitcoinPriceWorker.start();
    }
    
    @Override
    protected void onResume() {
        super.onResume();
        // Refresh the basket summary when returning to this activity
        updateBasketSummary();
    }
    
    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            finish();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }
    
    private void updateEmptyViewVisibility() {
        if (adapter.getItemCount() == 0) {
            emptyView.setVisibility(View.VISIBLE);
            recyclerView.setVisibility(View.GONE);
        } else {
            emptyView.setVisibility(View.GONE);
            recyclerView.setVisibility(View.VISIBLE);
        }
    }
    
    private void updateBasketSummary() {
        int itemCount = basketManager.getTotalItemCount();
        double totalPrice = basketManager.getTotalPrice();
        
        basketCountView.setText(String.valueOf(itemCount));
        
        // Format price with currency symbol
        String formattedTotal;
        if (itemCount > 0) {
            String currencySymbol = CurrencyManager.getInstance(this).getCurrentSymbol();
            formattedTotal = String.format("%s%.2f", currencySymbol, totalPrice);
            checkoutButton.setEnabled(true);
        } else {
            formattedTotal = "0.00";
            checkoutButton.setEnabled(false);
        }
        
        basketTotalView.setText(formattedTotal);
    }
    
    private void proceedToCheckout() {
        if (basketManager.getTotalItemCount() == 0) {
            Toast.makeText(this, "Your basket is empty", Toast.LENGTH_SHORT).show();
            return;
        }
        
        // Calculate the total amount in satoshis
        long satoshisAmount = basketManager.getTotalSatoshis(bitcoinPriceWorker.getCurrentPrice());
        
        // Create intent to go to ModernPOSActivity with the amount
        Intent intent = new Intent(this, ModernPOSActivity.class);
        
        // Add the satoshi amount as an extra (the ModernPOSActivity will need to be modified to accept this)
        intent.putExtra("EXTRA_PAYMENT_AMOUNT", satoshisAmount);
        
        // Don't clear the activity stack to allow returning to the item selection after payment
        
        // Clear the basket after checkout
        basketManager.clearBasket();
        
        // Start the activity
        startActivity(intent);
    }
    
    private class ItemAdapter extends RecyclerView.Adapter<ItemAdapter.ItemViewHolder> {
        
        private List<Item> items;
        private final Map<String, Integer> basketQuantities = new HashMap<>();
        
        ItemAdapter(List<Item> items) {
            this.items = items;
            // Initialize basket quantities
            for (BasketItem basketItem : basketManager.getBasketItems()) {
                basketQuantities.put(basketItem.getItem().getId(), basketItem.getQuantity());
            }
        }
        
        void updateItems(List<Item> newItems) {
            this.items = newItems;
            notifyDataSetChanged();
        }
        
        @NonNull
        @Override
        public ItemViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            View view = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.item_product_selection, parent, false);
            return new ItemViewHolder(view);
        }
        
        @Override
        public void onBindViewHolder(@NonNull ItemViewHolder holder, int position) {
            Item item = items.get(position);
            
            // Get quantity from map, default to 0
            int quantity = basketQuantities.getOrDefault(item.getId(), 0);
            
            holder.bind(item, quantity);
        }
        
        @Override
        public int getItemCount() {
            return items.size();
        }
        
        class ItemViewHolder extends RecyclerView.ViewHolder {
            private final TextView nameView;
            private final TextView variationView;
            private final TextView descriptionView;
            private final TextView skuView;
            private final TextView priceView;
            private final TextView quantityView;
            private final ImageButton decreaseButton;
            private final TextView basketQuantityView;
            private final ImageButton increaseButton;
            private final ImageView itemImageView;
            
            ItemViewHolder(@NonNull View itemView) {
                super(itemView);
                nameView = itemView.findViewById(R.id.item_name);
                variationView = itemView.findViewById(R.id.item_variation);
                descriptionView = itemView.findViewById(R.id.item_description);
                skuView = itemView.findViewById(R.id.item_sku);
                priceView = itemView.findViewById(R.id.item_price);
                quantityView = itemView.findViewById(R.id.item_quantity);
                decreaseButton = itemView.findViewById(R.id.decrease_quantity_button);
                basketQuantityView = itemView.findViewById(R.id.basket_quantity);
                increaseButton = itemView.findViewById(R.id.increase_quantity_button);
                itemImageView = itemView.findViewById(R.id.item_image);
            }
            
            void bind(Item item, int basketQuantity) {
                nameView.setText(item.getName());
                
                // Show variation if available
                if (item.getVariationName() != null && !item.getVariationName().isEmpty()) {
                    variationView.setVisibility(View.VISIBLE);
                    variationView.setText(item.getVariationName());
                } else {
                    variationView.setVisibility(View.GONE);
                }
                
                // Show description if available
                if (item.getDescription() != null && !item.getDescription().isEmpty()) {
                    descriptionView.setVisibility(View.VISIBLE);
                    descriptionView.setText(item.getDescription());
                } else {
                    descriptionView.setVisibility(View.GONE);
                }
                
                // Show SKU if available
                if (item.getSku() != null && !item.getSku().isEmpty()) {
                    skuView.setText("SKU: " + item.getSku());
                } else {
                    skuView.setText("");
                }
                
                // Format price with currency symbol
                String currencySymbol = CurrencyManager.getInstance(itemView.getContext()).getCurrentSymbol();
                priceView.setText(String.format("%s%.2f", currencySymbol, item.getPrice()));
                
                // Show quantity
                quantityView.setText("In stock: " + item.getQuantity());
                
                // Show basket quantity
                basketQuantityView.setText(String.valueOf(basketQuantity));
                
                // Load and display the item image
                if (item.getImagePath() != null && !item.getImagePath().isEmpty()) {
                    // Create a file from the image path
                    File imageFile = new File(item.getImagePath());
                    if (imageFile.exists()) {
                        // Load the image using Bitmap
                        Bitmap bitmap = BitmapFactory.decodeFile(imageFile.getAbsolutePath());
                        if (bitmap != null) {
                            itemImageView.setImageBitmap(bitmap);
                        } else {
                            // If bitmap loading fails, show placeholder
                            itemImageView.setImageResource(R.drawable.ic_image_placeholder);
                        }
                    } else {
                        // If file doesn't exist, show placeholder
                        itemImageView.setImageResource(R.drawable.ic_image_placeholder);
                    }
                } else {
                    // If no image path, show placeholder
                    itemImageView.setImageResource(R.drawable.ic_image_placeholder);
                }
                
                // Enable/disable decrease button based on basket quantity
                decreaseButton.setEnabled(basketQuantity > 0);
                
                // Enable/disable increase button based on stock
                boolean hasStock = item.getQuantity() > basketQuantity || item.getQuantity() == 0;
                increaseButton.setEnabled(hasStock);
                
                // Set up decrease button
                decreaseButton.setOnClickListener(v -> {
                    if (basketQuantity > 0) {
                        updateBasketItem(item, basketQuantity - 1);
                    }
                });
                
                // Set up increase button
                increaseButton.setOnClickListener(v -> {
                    // Check if we have stock available or if stock tracking is disabled (quantity = 0)
                    if (hasStock) {
                        updateBasketItem(item, basketQuantity + 1);
                    } else {
                        Toast.makeText(itemView.getContext(), "No more stock available", Toast.LENGTH_SHORT).show();
                    }
                });
                
                // Set up item click listener
                itemView.setOnClickListener(v -> {
                    // If we have stock available or stock tracking is disabled, add one item
                    if (hasStock) {
                        updateBasketItem(item, basketQuantity + 1);
                    } else {
                        Toast.makeText(itemView.getContext(), "No more stock available", Toast.LENGTH_SHORT).show();
                    }
                });
            }
            
            private void updateBasketItem(Item item, int newQuantity) {
                if (newQuantity <= 0) {
                    // Remove item from basket
                    basketManager.removeItem(item.getId());
                    basketQuantities.remove(item.getId());
                } else {
                    // Try to update item quantity first
                    boolean updated = basketManager.updateItemQuantity(item.getId(), newQuantity);
                    
                    // If update failed, item doesn't exist in basket yet, so add it
                    if (!updated) {
                        basketManager.addItem(item, newQuantity);
                    }
                    
                    basketQuantities.put(item.getId(), newQuantity);
                    
                    // Show a toast when adding an item
                    if (newQuantity == 1) {
                        Toast.makeText(itemView.getContext(), 
                                item.getDisplayName() + " added to basket", 
                                Toast.LENGTH_SHORT).show();
                    }
                }
                
                // Update UI
                notifyItemChanged(getAdapterPosition());
                updateBasketSummary();
            }
        }
    }
}

File: app/src/main/java/com/electricdreams/shellshock/feature/items/ItemSelectionActivity.kt
package com.electricdreams.shellshock.feature.items

import android.animation.AnimatorSet
import android.animation.ObjectAnimator
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.DecelerateInterpolator
import android.widget.Button
import android.widget.EditText
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.electricdreams.shellshock.PaymentRequestActivity
import com.electricdreams.shellshock.R
import com.electricdreams.shellshock.core.model.Amount
import com.electricdreams.shellshock.core.model.BasketItem
import com.electricdreams.shellshock.core.model.Item
import com.electricdreams.shellshock.core.util.BasketManager
import com.electricdreams.shellshock.core.util.CurrencyManager
import com.electricdreams.shellshock.core.util.ItemManager
import com.electricdreams.shellshock.core.worker.BitcoinPriceWorker
import java.io.File

class ItemSelectionActivity : AppCompatActivity() {

    private lateinit var itemManager: ItemManager
    private lateinit var basketManager: BasketManager
    private lateinit var bitcoinPriceWorker: BitcoinPriceWorker
    private lateinit var currencyManager: CurrencyManager

    // Views
    private lateinit var searchInput: EditText
    private lateinit var scanButton: ImageButton
    private lateinit var basketSection: LinearLayout
    private lateinit var basketRecyclerView: RecyclerView
    private lateinit var basketTotalView: TextView
    private lateinit var clearBasketButton: TextView
    private lateinit var itemsRecyclerView: RecyclerView
    private lateinit var emptyView: LinearLayout
    private lateinit var noResultsView: LinearLayout
    private lateinit var checkoutContainer: CardView
    private lateinit var checkoutButton: Button

    private lateinit var itemsAdapter: ItemsAdapter
    private lateinit var basketAdapter: BasketAdapter

    private var allItems: List<Item> = emptyList()
    private var filteredItems: List<Item> = emptyList()

    private val scannerLauncher = registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) { result ->
        if (result.resultCode == CheckoutScannerActivity.RESULT_BASKET_UPDATED) {
            // Refresh basket and items
            refreshBasket()
            itemsAdapter.notifyDataSetChanged()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_item_selection)

        itemManager = ItemManager.getInstance(this)
        basketManager = BasketManager.getInstance()
        bitcoinPriceWorker = BitcoinPriceWorker.getInstance(this)
        currencyManager = CurrencyManager.getInstance(this)

        initViews()
        setupListeners()
        setupRecyclerViews()

        loadItems()
        refreshBasket()

        bitcoinPriceWorker.start()
    }

    private fun initViews() {
        findViewById<View>(R.id.back_button).setOnClickListener { finish() }

        searchInput = findViewById(R.id.search_input)
        scanButton = findViewById(R.id.scan_button)
        basketSection = findViewById(R.id.basket_section)
        basketRecyclerView = findViewById(R.id.basket_recycler_view)
        basketTotalView = findViewById(R.id.basket_total)
        clearBasketButton = findViewById(R.id.clear_basket_button)
        itemsRecyclerView = findViewById(R.id.items_recycler_view)
        emptyView = findViewById(R.id.empty_view)
        noResultsView = findViewById(R.id.no_results_view)
        checkoutContainer = findViewById(R.id.checkout_container)
        checkoutButton = findViewById(R.id.checkout_button)
    }

    private fun setupListeners() {
        scanButton.setOnClickListener {
            val intent = Intent(this, CheckoutScannerActivity::class.java)
            scannerLauncher.launch(intent)
        }

        searchInput.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
            override fun afterTextChanged(s: Editable?) {
                filterItems(s?.toString() ?: "")
            }
        })

        clearBasketButton.setOnClickListener {
            AlertDialog.Builder(this)
                .setTitle("Clear Basket")
                .setMessage("Remove all items from basket?")
                .setPositiveButton("Clear") { _, _ ->
                    basketManager.clearBasket()
                    itemsAdapter.clearAllQuantities()
                    refreshBasket()
                }
                .setNegativeButton("Cancel", null)
                .show()
        }

        checkoutButton.setOnClickListener {
            proceedToCheckout()
        }
    }

    private fun setupRecyclerViews() {
        // Items list
        itemsRecyclerView.layoutManager = LinearLayoutManager(this)
        itemsAdapter = ItemsAdapter()
        itemsRecyclerView.adapter = itemsAdapter

        // Basket list
        basketRecyclerView.layoutManager = LinearLayoutManager(this)
        basketAdapter = BasketAdapter()
        basketRecyclerView.adapter = basketAdapter
    }

    private fun loadItems() {
        allItems = itemManager.getAllItems()
        filteredItems = allItems
        itemsAdapter.updateItems(filteredItems)
        updateEmptyState()
    }

    private fun filterItems(query: String) {
        filteredItems = if (query.isBlank()) {
            allItems
        } else {
            itemManager.searchItems(query)
        }
        itemsAdapter.updateItems(filteredItems)
        updateEmptyState()
    }

    private fun updateEmptyState() {
        val hasItems = allItems.isNotEmpty()
        val hasResults = filteredItems.isNotEmpty()
        val isSearching = searchInput.text.isNotBlank()

        when {
            !hasItems -> {
                emptyView.visibility = View.VISIBLE
                noResultsView.visibility = View.GONE
                itemsRecyclerView.visibility = View.GONE
            }
            !hasResults && isSearching -> {
                emptyView.visibility = View.GONE
                noResultsView.visibility = View.VISIBLE
                itemsRecyclerView.visibility = View.GONE
            }
            else -> {
                emptyView.visibility = View.GONE
                noResultsView.visibility = View.GONE
                itemsRecyclerView.visibility = View.VISIBLE
            }
        }
    }

    private fun refreshBasket() {
        val basketItems = basketManager.getBasketItems()
        basketAdapter.updateItems(basketItems)

        if (basketItems.isEmpty()) {
            // Hide basket section with animation
            if (basketSection.visibility == View.VISIBLE) {
                animateBasketSection(false)
            }
            checkoutContainer.visibility = View.GONE
        } else {
            // Show basket section with animation
            if (basketSection.visibility != View.VISIBLE) {
                animateBasketSection(true)
            }
            checkoutContainer.visibility = View.VISIBLE
            updateCheckoutButton()
        }

        updateBasketTotal()
    }

    private fun animateBasketSection(show: Boolean) {
        if (show) {
            basketSection.visibility = View.VISIBLE
            basketSection.alpha = 0f
            basketSection.translationY = -50f

            AnimatorSet().apply {
                playTogether(
                    ObjectAnimator.ofFloat(basketSection, "alpha", 0f, 1f),
                    ObjectAnimator.ofFloat(basketSection, "translationY", -50f, 0f)
                )
                duration = 300
                interpolator = DecelerateInterpolator()
                start()
            }
        } else {
            val animSet = AnimatorSet().apply {
                playTogether(
                    ObjectAnimator.ofFloat(basketSection, "alpha", 1f, 0f),
                    ObjectAnimator.ofFloat(basketSection, "translationY", 0f, -50f)
                )
                duration = 200
                interpolator = DecelerateInterpolator()
            }
            animSet.addListener(object : android.animation.AnimatorListenerAdapter() {
                override fun onAnimationEnd(animation: android.animation.Animator) {
                    basketSection.visibility = View.GONE
                }
            })
            animSet.start()
        }
    }

    private fun updateBasketTotal() {
        val itemCount = basketManager.getTotalItemCount()
        val fiatTotal = basketManager.getTotalPrice()
        val satsTotal = basketManager.getTotalSatsDirectPrice()

        val formattedTotal = if (itemCount > 0) {
            val currencyCode = currencyManager.getCurrentCurrency()
            val currency = Amount.Currency.fromCode(currencyCode)

            when {
                fiatTotal > 0 && satsTotal > 0 -> {
                    val fiatAmount = Amount.fromMajorUnits(fiatTotal, currency)
                    "$fiatAmount + $satsTotal sats"
                }
                satsTotal > 0 -> "$satsTotal sats"
                else -> Amount.fromMajorUnits(fiatTotal, currency).toString()
            }
        } else {
            "0.00"
        }

        basketTotalView.text = formattedTotal
    }

    private fun updateCheckoutButton() {
        val fiatTotal = basketManager.getTotalPrice()
        val satsTotal = basketManager.getTotalSatsDirectPrice()
        val currencyCode = currencyManager.getCurrentCurrency()
        val currency = Amount.Currency.fromCode(currencyCode)

        val buttonText = when {
            fiatTotal > 0 && satsTotal > 0 -> {
                val fiatAmount = Amount.fromMajorUnits(fiatTotal, currency)
                "Charge $fiatAmount + $satsTotal sats"
            }
            satsTotal > 0 -> "Charge $satsTotal sats"
            else -> {
                val fiatAmount = Amount.fromMajorUnits(fiatTotal, currency)
                "Charge $fiatAmount"
            }
        }

        checkoutButton.text = buttonText
    }

    private fun proceedToCheckout() {
        if (basketManager.getTotalItemCount() == 0) {
            Toast.makeText(this, "Your basket is empty", Toast.LENGTH_SHORT).show()
            return
        }

        val fiatTotal = basketManager.getTotalPrice()
        val satsTotal = basketManager.getTotalSatsDirectPrice()
        val btcPrice = bitcoinPriceWorker.getCurrentPrice()

        // Calculate total in satoshis
        val totalSatoshis = basketManager.getTotalSatoshis(btcPrice)

        if (totalSatoshis <= 0) {
            Toast.makeText(this, "Invalid payment amount", Toast.LENGTH_SHORT).show()
            return
        }

        // Determine how to format the amount for PaymentRequestActivity
        val formattedAmount: String
        
        when {
            // Pure fiat (no sats items) - display as fiat
            satsTotal == 0L && fiatTotal > 0 -> {
                val currencyCode = currencyManager.getCurrentCurrency()
                val currency = Amount.Currency.fromCode(currencyCode)
                // Convert fiat total to cents for Amount class
                val fiatCents = (fiatTotal * 100).toLong()
                formattedAmount = Amount(fiatCents, currency).toString()
            }
            // Pure sats (no fiat items) - display as BTC/sats
            fiatTotal == 0.0 && satsTotal > 0 -> {
                formattedAmount = Amount(satsTotal, Amount.Currency.BTC).toString()
            }
            // Mixed fiat + sats - treat as pure sats (display as BTC)
            else -> {
                formattedAmount = Amount(totalSatoshis, Amount.Currency.BTC).toString()
            }
        }

        val intent = Intent(this, PaymentRequestActivity::class.java).apply {
            putExtra(PaymentRequestActivity.EXTRA_PAYMENT_AMOUNT, totalSatoshis)
            putExtra(PaymentRequestActivity.EXTRA_FORMATTED_AMOUNT, formattedAmount)
        }

        basketManager.clearBasket()
        startActivity(intent)
        finish()
    }

    override fun onResume() {
        super.onResume()
        // Refresh in case items were modified
        loadItems()
        refreshBasket()
    }

    // ==================== Items Adapter ====================

    private inner class ItemsAdapter : RecyclerView.Adapter<ItemsAdapter.ItemViewHolder>() {

        private var items: List<Item> = emptyList()
        private val basketQuantities: MutableMap<String, Int> = mutableMapOf()

        fun updateItems(newItems: List<Item>) {
            items = newItems
            refreshBasketQuantities()
            notifyDataSetChanged()
        }

        private fun refreshBasketQuantities() {
            basketQuantities.clear()
            for (basketItem in basketManager.getBasketItems()) {
                basketQuantities[basketItem.item.id!!] = basketItem.quantity
            }
        }

        fun clearAllQuantities() {
            basketQuantities.clear()
            notifyDataSetChanged()
        }

        fun resetItemQuantity(itemId: String) {
            basketQuantities.remove(itemId)
            notifyDataSetChanged()
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ItemViewHolder {
            val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.item_product_selection, parent, false)
            return ItemViewHolder(view)
        }

        override fun onBindViewHolder(holder: ItemViewHolder, position: Int) {
            val item = items[position]
            val quantity = basketQuantities[item.id] ?: 0
            holder.bind(item, quantity, position == items.lastIndex)
        }

        override fun getItemCount(): Int = items.size

        inner class ItemViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
            private val nameView: TextView = itemView.findViewById(R.id.item_name)
            private val variationView: TextView = itemView.findViewById(R.id.item_variation)
            private val priceView: TextView = itemView.findViewById(R.id.item_price)
            private val stockView: TextView = itemView.findViewById(R.id.item_quantity)
            private val quantityView: TextView = itemView.findViewById(R.id.basket_quantity)
            private val decreaseButton: ImageButton = itemView.findViewById(R.id.decrease_quantity_button)
            private val increaseButton: ImageButton = itemView.findViewById(R.id.increase_quantity_button)
            private val itemImageView: ImageView = itemView.findViewById(R.id.item_image)
            private val imagePlaceholder: ImageView? = itemView.findViewById(R.id.item_image_placeholder)
            private val divider: View? = itemView.findViewById(R.id.divider)

            fun bind(item: Item, basketQuantity: Int, isLast: Boolean) {
                nameView.text = item.name ?: ""

                if (!item.variationName.isNullOrEmpty()) {
                    variationView.text = item.variationName
                    variationView.visibility = View.VISIBLE
                } else {
                    variationView.visibility = View.GONE
                }

                priceView.text = item.getFormattedPrice()

                if (item.trackInventory) {
                    stockView.visibility = View.VISIBLE
                    stockView.text = "${item.quantity} in stock"
                } else {
                    stockView.visibility = View.GONE
                }

                quantityView.text = basketQuantity.toString()

                // Load image
                loadItemImage(item)

                // Button states
                decreaseButton.isEnabled = basketQuantity > 0
                decreaseButton.alpha = if (basketQuantity > 0) 1f else 0.4f

                val hasStock = if (item.trackInventory) item.quantity > basketQuantity else true
                increaseButton.isEnabled = hasStock
                increaseButton.alpha = if (hasStock) 1f else 0.4f

                // Hide divider on last item
                divider?.visibility = if (isLast) View.GONE else View.VISIBLE

                decreaseButton.setOnClickListener {
                    if (basketQuantity > 0) {
                        updateBasketItem(item, basketQuantity - 1)
                    }
                }

                increaseButton.setOnClickListener {
                    if (hasStock) {
                        updateBasketItem(item, basketQuantity + 1)
                    } else {
                        Toast.makeText(itemView.context, "No more stock available", Toast.LENGTH_SHORT).show()
                    }
                }
            }

            private fun loadItemImage(item: Item) {
                if (!item.imagePath.isNullOrEmpty()) {
                    val imageFile = File(item.imagePath!!)
                    if (imageFile.exists()) {
                        val bitmap: Bitmap? = BitmapFactory.decodeFile(imageFile.absolutePath)
                        if (bitmap != null) {
                            itemImageView.setImageBitmap(bitmap)
                            imagePlaceholder?.visibility = View.GONE
                            return
                        }
                    }
                }
                itemImageView.setImageBitmap(null)
                imagePlaceholder?.visibility = View.VISIBLE
            }

            private fun updateBasketItem(item: Item, newQuantity: Int) {
                val wasEmpty = basketManager.getTotalItemCount() == 0

                if (newQuantity <= 0) {
                    basketManager.removeItem(item.id!!)
                    basketQuantities.remove(item.id!!)
                } else {
                    val updated = basketManager.updateItemQuantity(item.id!!, newQuantity)
                    if (!updated) {
                        basketManager.addItem(item, newQuantity)
                    }
                    basketQuantities[item.id!!] = newQuantity
                }

                // Animate quantity change
                quantityView.text = newQuantity.toString()
                val scaleAnim = AnimatorSet().apply {
                    playSequentially(
                        ObjectAnimator.ofFloat(quantityView, "scaleX", 1f, 1.2f).setDuration(100),
                        ObjectAnimator.ofFloat(quantityView, "scaleX", 1.2f, 1f).setDuration(100)
                    )
                }
                val scaleAnimY = AnimatorSet().apply {
                    playSequentially(
                        ObjectAnimator.ofFloat(quantityView, "scaleY", 1f, 1.2f).setDuration(100),
                        ObjectAnimator.ofFloat(quantityView, "scaleY", 1.2f, 1f).setDuration(100)
                    )
                }
                AnimatorSet().apply {
                    playTogether(scaleAnim, scaleAnimY)
                    start()
                }

                notifyItemChanged(adapterPosition)
                refreshBasket()
            }
        }
    }

    // ==================== Basket Adapter ====================

    private inner class BasketAdapter : RecyclerView.Adapter<BasketAdapter.BasketViewHolder>() {

        private var basketItems: List<BasketItem> = emptyList()

        fun updateItems(newItems: List<BasketItem>) {
            val oldSize = basketItems.size
            basketItems = newItems

            // Simple diff - animate new items
            if (newItems.size > oldSize) {
                notifyItemInserted(0) // Newest items at top
            } else {
                notifyDataSetChanged()
            }
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BasketViewHolder {
            val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.item_basket_compact, parent, false)
            return BasketViewHolder(view)
        }

        override fun onBindViewHolder(holder: BasketViewHolder, position: Int) {
            // Show newest first (reverse order)
            val basketItem = basketItems[basketItems.size - 1 - position]
            holder.bind(basketItem)
        }

        override fun getItemCount(): Int = basketItems.size

        inner class BasketViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
            private val quantityBadge: TextView = itemView.findViewById(R.id.item_quantity_badge)
            private val nameView: TextView = itemView.findViewById(R.id.item_name)
            private val variationView: TextView = itemView.findViewById(R.id.item_variation)
            private val totalView: TextView = itemView.findViewById(R.id.item_total)
            private val removeButton: ImageButton = itemView.findViewById(R.id.remove_button)

            fun bind(basketItem: BasketItem) {
                val item = basketItem.item

                quantityBadge.text = basketItem.quantity.toString()
                nameView.text = item.name ?: ""

                if (!item.variationName.isNullOrEmpty()) {
                    variationView.text = item.variationName
                    variationView.visibility = View.VISIBLE
                } else {
                    variationView.visibility = View.GONE
                }

                // Format total
                if (basketItem.isSatsPrice()) {
                    totalView.text = "${basketItem.getTotalSats()} sats"
                } else {
                    val currencyCode = currencyManager.getCurrentCurrency()
                    val currency = Amount.Currency.fromCode(currencyCode)
                    val totalAmount = Amount.fromMajorUnits(basketItem.getTotalPrice(), currency)
                    totalView.text = totalAmount.toString()
                }

                removeButton.setOnClickListener {
                    val itemId = item.id!!
                    basketManager.removeItem(itemId)
                    itemsAdapter.resetItemQuantity(itemId)
                    refreshBasket()
                }
            }
        }
    }
}
